### ISSUE TITLE
Weight initialization for Neural network

### ISSUE THREAD

**Issue by CDLuminate** [torch/nn#938]
I'm trying to train a model which combines CNN and LSTM together, however the LSTM weights initialized by default is some what bad -- I got a strange loss curve and some other strange observations. However all of these are fixed with LSTM weight re-initialization, just like this:

```lua
local model = Model()
local params, grad_params = model:getParameters()
params:copy(torch.randn(params:size())) -- This fixes the above issue
```

Besides, when we are going to train a CNN with caffe, explicit weight initialization method is required, e.g. 
```
# from bvlc_alexnet
343 layer {
344   name: "fc8"
345   type: "InnerProduct"
346   bottom: "fc7"
347   top: "fc8"
...
356   inner_product_param {
357     num_output: 1000
358     weight_filler {
359       type: "gaussian"
360       std: 0.01
361     }
362     bias_filler {
363       type: "constant"
364       value: 0
365     }
366   }
367 }
```
However in the torch's documentation the weight initialization problem is not mentioned in the documentation.

For example, the weight initialization method is fixed for linear module as follows,
```lua
# linear.lua
 21 function Linear:reset(stdv)                                                     
 22    if stdv then                                                                 
 23       stdv = stdv * math.sqrt(3)                                                
 24    else                                                                         
 25       stdv = 1./math.sqrt(self.weight:size(2))                                  
 26    end                                                                          
 27    if nn.oldSeed then                                                           
 28       for i=1,self.weight:size(1) do                                            
 29          self.weight:select(1, i):apply(function()                              
 30             return torch.uniform(-stdv, stdv)                                   
 31          end)                                                                   
 32       end                                                                       
 33       if self.bias then                                                         
 34          for i=1,self.bias:nElement() do                                        
 35             self.bias[i] = torch.uniform(-stdv, stdv)                           
 36          end                                                                    
 37       end                                                                       
 38    else                                                                         
 39       self.weight:uniform(-stdv, stdv)                                          
 40       if self.bias then self.bias:uniform(-stdv, stdv) end                      
 41    end                                                                          
 42    return self                                                                  
 43 end
```

Is torch confident on the default weight initialization method? I think there should be an elegant way to override the weight initialzation, instead of
```lua
params:copy(torch.randn(params:size())) 
```
Thanks.

**Comment by fmassa** [comment]
You could check https://github.com/Kaixhin/nninit for the weight init schemes

**Comment by Kaixhin** [comment]
The default comes from the [Efficient BackProp](https://groups.google.com/forum/#!topic/digits-users/QmhMM8OrQK8) paper, and is basically the Kaiming/He initialisation method. But yes, [`nninit`](https://github.com/Kaixhin/nninit) was made to deal with weight initialisations, and can be adapted to LSTMs (check the example code).

**Comment by CDLuminate** [comment]
@fmassa @Kaixhin Thank you! `nninit` seems to be very helpful to me.

**Comment by Kaixhin** [comment]
@CDLuminate If `nninit` suits your needs could you please close this issue? Thanks.
